nilearn/decoding/space_net_solvers.py

```diff
@@ -0,0 +1,533 @@
+"""
+Regression with spatial priors like TV-L1 and Smooth LASSO.
+
+"""
+# Author: DOHMATOB Elvis Dopgima,
+#         Gael Varoquaux,
+#         Alexandre Gramfort,
+#         Gaspar Pizarro,
+#         Virgile Fritsch,
+#         Bertrand Thirion,
+#         and others.
+# License: simplified BSD
+
+from math import sqrt
+import numpy as np
+from .objective_functions import (spectral_norm_squared,
+                                  gradient_id,
+                                  logistic_loss_lipschitz_constant,
+                                  squared_loss, squared_loss_grad, _unmask,
+                                  logistic_loss_grad,
+                                  logistic as logistic_loss)
+from .objective_functions import gradient, div
+from .proximal_operators import (_prox_l1, _prox_l1_with_intercept,
+                                 _prox_tvl1, _prox_tvl1_with_intercept)
+from .fista import mfista
+
+
+def squared_loss_and_spatial_grad(X, y, w, mask, grad_weight):
+    """
+    Computes the differentiable loss function (squared_loss + grad_weight
+    * gradient) for Smooth-Lasso regularization
+
+    Parameters
+    ----------
+    X : ndarray, shape (n_samples, n_features)
+        Design matrix.
+
+    y : ndarray, shape (n_samples,)
+        Target / response vector.
+
+    w : ndarray shape (n_features,)
+        Unmasked, ravelized weights map.
+
+    grad_weight: float
+        l1_ratio * alpha
+
+    Returns
+    -------
+    float
```

Comment:
Please clarify: If I understand correctly, this is the squared l2 norm of the gradient.